lib/semantics/expression.cc

```diff
@@ -2540,6 +2549,43 @@ MaybeExpr ArgumentAnalyzer::TryDefinedOp(
   return TryDefinedOp(oprs[0], std::move(error));
 }
 
+std::optional<ProcedureRef> ArgumentAnalyzer::TryDefinedAssignment() {
+  using semantics::Tristate;
+  const auto &lhs{GetAsExpr(0)};
+  const auto &rhs{GetAsExpr(1)};
```

Comment:
Isn't `GetAsExpr` returning values ? I think it might be better to change `GetAsExpr` so it would return `Expr<SomeExpr>&` and so that the expression can also be moved instead of being copied after the  analysis of intrinsic assignments at:

https://github.com/flang-compiler/f18/blob/35a31138f08ff2d41e3fb016f3e33bfde4cd835e/lib/semantics/expression.cc#L1801

Reply:
OK, done.

Reply:
The name should probably be changed to reflect the altered semantics so that there's less risk of failing to consider the lifetime of the referent.

Reply:
OK, I made it `MoveExpr`.

Reply:
Now it returns an rvalue reference to the expression, which remains in place in the actual argument structure; the expression is not actually moved, despite the new `MoveExpr` name.

If you'd like to leave it the expression in place, I think you should grab a `const &` reference to it via some API whose name indicates that a reference has been created.  If you want to get a copy of the expression, the original API did that.

Reply:
I don't see where the the result of `MoveExpr` is being copied now. What would I have to do to fix it?

If I use just `GetAsExpr`, returning `const Expr<SomeType> &`, then in each of the places I need an r-value reference (`BinaryLogicalOperation`, `Relate`, etc.) I'll need to use `common::Clone` or something to make a copy. That doesn't seem like a good approach.

Reply:
Then it might be more simple to revert `GetAsExpr` to be the point that makes the copy.